server/src/main/java/org/elasticsearch/search/lookup/SourceLookup.java

```diff
@@ -62,101 +58,32 @@ public int docId() {
-            fieldReader.accept(docId, sourceFieldVisitor);
-            BytesReference source = sourceFieldVisitor.source();
-            if (source == null) {
-                this.source = emptyMap();
-                this.sourceContentType = null;
-            } else {
-                Tuple<XContentType, Map<String, Object>> tuple = sourceAsMapAndType(source);
-                this.sourceContentType = tuple.v1();
-                this.source = tuple.v2();
-            }
-        } catch (Exception e) {
-            throw new ElasticsearchParseException("failed to parse / load source", e);
-        }
-        return this.source;
+        return sourceProvider.source();
     }
 
-    private static Tuple<XContentType, Map<String, Object>> sourceAsMapAndType(BytesReference source) throws ElasticsearchParseException {
-        return XContentHelper.convertToMap(source, false);
-    }
-
-    /**
-     * Get the source as a {@link Map} of java objects.
-     * <p>
-     * Important: This can lose precision on numbers with a decimal point. It
-     * converts numbers like {@code "n": 1234.567} to a {@code double} which
-     * only has 52 bits of precision in the mantissa. This will come up most
-     * frequently when folks write nanosecond precision dates as a decimal
-     * number.
-     */
-    public static Map<String, Object> sourceAsMap(BytesReference source) throws ElasticsearchParseException {
-        return sourceAsMapAndType(source).v2();
+    public void setSourceProvider(SourceProvider sourceProvider) {
+        this.sourceProvider = sourceProvider;
     }
 
     public void setSegmentAndDocument(LeafReaderContext context, int docId) {
-        // if we are called with the same document, don't invalidate source
-        if (this.reader == context.reader() && this.docId == docId) {
-            return;
-        }
-
-        // only reset reader and fieldReader when reader changes
-        if (this.reader != context.reader()) {
-            this.reader = context.reader();
-
-            // All the docs to fetch are adjacent but Lucene stored fields are optimized
-            // for random access and don't optimize for sequential access - except for merging.
-            // So we do a little hack here and pretend we're going to do merges in order to
-            // get better sequential access.
-            if (context.reader()instanceof SequentialStoredFieldsLeafReader lf) {
-                // Avoid eagerly loading the stored fields reader, since this can be expensive
-                Supplier<StoredFieldsReader> supplier = new MemoizedSupplier<>(lf::getSequentialStoredFieldsReader);
-                fieldReader = (d, v) -> supplier.get().visitDocument(d, v);
-            } else {
-                fieldReader = context.reader()::document;
-            }
+        if (sourceProvider instanceof ReaderSourceProvider) {
+            ((ReaderSourceProvider) sourceProvider).setSegmentAndDocument(context, docId);
         }
```

Comment:
Let's add a `setSegmentAndDocument()` method to the `SourceProvider` interface as well.